design/20230302.gomod.md

```diff
@@ -0,0 +1,244 @@
+Since we have one `go.mod` file for all of our built binaries, it's not possible for us to be selective about upgrades,
+either.  If, say, only the `controller` component were to report as having a critical vulnerability, we'd have no
+way of fixing only that one vulnerability while leaving everything else untouched.
+
+Essentially, our current project layout forces us to made difficult choices whenever we need to upgrade things.
+
+### Problem Example
+
+**In short:** An example of how upgrades can be particularly difficult in some cases, with no good options.
+
+At the time of writing, cert-manager 1.10 is still in support and depends on Helm because it's imported by `cmctl` (and
+only `cmctl`). We can see the dependency in [go.mod](https://github.com/cert-manager/cert-manager/blob/f54dd1dc98900607e1db7bd4ac2512f0bfe39301/go.mod#L41).
+
+There's a vulnerability reported for Helm v3.10.3 ([1]) which we'd like to patch, but the only version with a fix
+available is Helm v3.11.1.
+
+Between Helm 3.10 and 3.11, several of Helm's dependencies were upgraded, and crucially Helm has some of the same
+dependencies that cert-manager does. That means that we can't easily _just_ upgrade Helm.
+
+Running `go get -u helm.sh/helm/v3` produces 56 different upgrades of _other_ dependencies. Notably, it bumps our
+Kubernetes dependencies from v0.25.2 to v0.26.0 but there are several other changes.
+
+(NB: Helm is just an example here and we could have problems with any package)
+
+## Proposed Solution: Go Module Proliferation
+
+**In short:** Add several new `go.mod` files so individual components can be patched independently
+
+We can create several new Go modules so that each binary we build can have distinct dependencies. This would mean that
+`cmctl` having a dependency on Helm would only affect `cmctl` and wouldn't force us to change any of the other
+components we build in order to patch a Helm vulnerability.
+
+In addition to separating out the binaries, we could also separate out the "core" issuer implementations into a
+distinct module (since they bring so many external dependencies and currently live in `pkg/` where they'll be part of
+the core cert-manager `go.mod` file).
+
+Plus, where we have testing-only dependencies (e.g. for integration or end-to-end tests) we could create a test module
+so that those test dependencies don't pollute the main `go.mod`.
+
+## Potential Issues
+
+### Importing cert-manager / Development Experience
+
+**In short:** Module replacements help developers but don't aren't respected by imports, which means some changes could need two PRs
+
+**Useful Reference:** It helps to read [this StackOverflow comment](https://stackoverflow.com/a/71984158) to better understand the options we have
+
+The simplest development experience when working with multiple Go modules at once is to use either the `replace` directive
+in `go.mod` `use` in `go.work` to point to local versions of a module. This allows both modules to be developed in
```

Comment:
I don't think you need `go.work` if you have a replace statement right?

Reply:
That's correct; go.work is an easier way to manage local development in the case where you _don't_ use replaces everywhere.